### ISSUE TITLE
Fatal Exception in Express processing with CMSSW_13_0_7

### ISSUE THREAD

**Issue by francescobrivio** [cms-sw/cmssw#41843]
There are some reported crashed in the Express processing of `StreamHLTMonitor` for run `368318` (see [CMSTalk post](https://cms-talk.web.cern.ch/t/paused-express-jobs-for-run-368318-hltmonitor-with-invalidreference-error/24927)).

**The exception is:**
```
An exception of category 'InvalidReference' occurred while
   [0] Processing  Event run: 368318 lumi: 159 event: 237000809 stream: 0
   [1] Running path 'dqmoffline_step'
   [2] Calling method for module ParticleNetJetTagMonitor/'particleNetAK8HbbTagMonitoring'
Exception Message:
attempting to get view from an unavailable RefToBaseProd.
```

**Recipe to reproduce the error:**
```
cmsrel CMSSW_13_0_7
cd CMSSW_13_0_7/src
cmsenv
cp /eos/cms/store/logs/prod/recent/Express/Express_Run368318_StreamHLTMonitor/Express/vocms0314.cern.ch-441154-0-log.tar.gz .
tar -zxvf vocms0314.cern.ch-441154-0-log.tar.gz 
cd job/WMTaskSpace/cmsRun1
```
edit the `PSet.py` to be:
```
import FWCore.ParameterSet.Config as cms
import pickle
with open('PSet.pkl', 'rb') as handle:
    process = pickle.load(handle)
    process.options.numberOfThreads=cms.untracked.uint32(1)
    process.options.numberOfStreams=cms.untracked.uint32(1)
    process.source.skipEvents = cms.untracked.uint32(169)
```
run it with:
```
cmsRun PSet.py
```

**Comment by francescobrivio** [comment]
assign reconstruction,btv-pog

**Comment by francescobrivio** [comment]
A possible PR which entered in 13_0_7 and touched PNet DQM is https://github.com/cms-sw/cmssw/pull/41704

**Comment by francescobrivio** [comment]
assign dqm

**Comment by francescobrivio** [comment]
> A possible PR which entered in 13_0_7 and touched PNet DQM is #41704

let me tag also explicitly the people involved in this: @marinakolosova @rgerosa @scooperstein

**Comment by francescobrivio** [comment]
assign hlt

 - maybe the issue si coming from the menu? @trtomei @mzarucki

**Comment by francescobrivio** [comment]
It might be a misconfiguration of the eventContent of the `HLTMonitor` stream (not yet confirmed).
@trtomei is testing it now on Hilton.

**Comment by missirol** [comment]
I think the EventContent of the HLTMonitor stream was incorrect, and has been fixed in [CMSHLT-2825](https://its.cern.ch/jira/browse/CMSHLT-2825). I think we have seen it only after deploying `13_0_7` at T0 (as opposed to, after deploying the HLT menu with the incorrect EventContent) because #41704 exposed this mistake.

That said, I haven't verified that this fix (which is correct anyway) will solve this issue. If it does, the problem should not occur anymore starting from run-368321.

**Comment by francescobrivio** [comment]
Indeed starting from run `368321` we don't see anymore failed jobs.
I'll keep monitoring the processing and if all goes fine I'll close the issue.

**Comment by francescobrivio** [comment]
> Indeed starting from run 368321 we don't see anymore failed jobs.
> I'll keep monitoring the processing and if all goes fine I'll close the issue.

Ok I haven't seen any other failure of this kind after run `368321` (when the new menu was deployed).

Before closing this issue, is there something we could do to:
 - avoid crashing the whole job? (some protection/check before the dereferenciation?)
 - add a test for suche cases?

**Comment by makortel** [comment]
> * avoid crashing the whole job? (some protection/check before the dereferenciation?)

In principle adding something along the following
```cpp
if (not jetPNETScoreHLTHandle->keyProduct().isAvailable()) {
      edm::LogWarning("ParticleNetJetTagMonitor") << "Collection used as a key by HLT Jet tags collection is not available, will skip event";
      return;
}
```
before the loop
https://github.com/cms-sw/cmssw/blob/2625240908ff0eb2ef39fb795d76c8a1a9b91973/DQMOffline/Trigger/plugins/ParticleNetJetTagMonitor.cc#L846
would the appropriate check for the availability of the key collection. Unfortunately that code does not work today, because `edm::RefToBaseProd` (the type returned by `keyProduct()` above) does not have `isAvailable()` member function. It looks to be straightforward to add though (and therefore will do).

**Comment by makortel** [comment]
> * add a test for suche cases?

The framework provides a TestProcessor facility that enables fairly easy testing of modules in isolation e.g. for various corner cases such as "Run had no lumis", "Lumi had no events", "Event does not contain the necessary data products". It is documented in https://github.com/cms-sw/cmssw/blob/master/FWCore/TestProcessor/Readme.md, and many examples can be found with "git grep TestProcessor". A notable example is in the skeleton templates https://github.com/cms-sw/cmssw/blob/master/FWCore/Skeletons/mkTemplates/EDProducer/test_catch2_EDProducer.cc

**Comment by makortel** [comment]
> Unfortunately that code does not work today, because `edm::RefToBaseProd` (the type returned by `keyProduct()` above) does not have `isAvailable()` member function. It looks to be straightforward to add though (and therefore will do).

The `isAvailable()` member function is added to `edm::RefToBaseProd` in https://github.com/cms-sw/cmssw/pull/41858.

**Comment by mmusich** [comment]
> The `isAvailable()` member function is added to `edm::RefToBaseProd `in https://github.com/cms-sw/cmssw/pull/41858.

given the thread in https://github.com/cms-sw/cmssw/pull/41858#discussion_r1221847828 it's not clear to me if the recommendation is to NOT protect the DQM code with this because of disk reading costs, or not.

**Comment by makortel** [comment]
> > The `isAvailable()` member function is added to `edm::RefToBaseProd `in #41858.
> 
> given the thread in [#41858 (comment)](https://github.com/cms-sw/cmssw/pull/41858#discussion_r1221847828) it's not clear to me if the recommendation is to NOT protect the DQM code with this because of disk reading costs, or not.

The DQM code de-references the corresponding `RefToBase`'s in
https://github.com/cms-sw/cmssw/blob/d2f6636f4568e620182fe0782133a243bfacf4fb/DQMOffline/Trigger/plugins/ParticleNetJetTagMonitor.cc#L871-L872
so the data would be read from storage in any case. So the cost of an `isAvailable()` check there should be negligible.

**Comment by francescobrivio** [comment]
I'll close this given it has been addressed in #41858 + #41930. Thanks everyone!

**Comment by missirol** [comment]
+hlt

The EventContent of the HLTMonitor stream was fixed in [CMSHLT-2825](https://its.cern.ch/jira/browse/CMSHLT-2825).
